docs/Planner.md

```diff
@@ -1,63 +1,123 @@
-# Planner Design
+# Query Planner and Optimizer
 
-This document lays out the design of the planner.
+This document describes the design of the Flux query planner and optimizer.
 
-## Interface
+## Plans
+--------
+
+A Flux query plan is represented as a Directed Acyclic Graph (DAG).
+The nodes of the graph represent individual query operations to perform.
+The edges of the graph represent the flow of data from one operation to another.
+In a Flux query plan, data flows upwards from source nodes to terminal yield nodes.
 
-The Planner inter is defined as:
+## Plan Nodes
+-------------
 
 ```go
-type Planner interface {
-    Plan(Query, []Storage) Plan
-}
-```
+type PlanNode interface {
+	// Returns an identifier for this plan node
+	ID() NodeID
 
-The planner is responsible for taking a query DAG and a set of available storage interfaces and produce a plan DAG.
+	// Returns the time bounds for this plan node
+	Bounds() *Bounds
 
-## Plans
+	// Plan nodes executed immediately before this node
+	Predecessors() []PlanNode
 
-Plans are created via a two step process:
+	// Plan nodes executed immediately after this node
+	Successors() []PlanNode
 
-1. Create a general plan from the query.
-2. Create a specific plan from the general plan and available storage interface.
+	// Specification of the procedure represented by this node
+	ProcedureSpec() ProcedureSpec
 
-The general plan specifies all the needed data frames and their lineage needed to produce the final query result.
-The specific plan specifies how the general plan will be executed, which storage interfaces will be consumed and how.
+	// Type of procedure represented by this node
+	Kind() ProcedureKind
 
-The general plan does not leave the scope of the Planner and is not part of the API of the planner.
-Hence the Plan type above it the specific plan.
+    ...
+}
+```
 
-## Plan DAG
+The nodes of a query represent operations.
+An operation is a transformation to apply to data in some form.
+Nodes have predecessors and successors.
+A node's predecessors are the query operations executed immediately before it.
+A node's successors are the query operations executed immediately after it.
+A node receives data directly from its predecessors.
+A node feeds data directly to its successors.
 
-Both the general and specific plans are represented as DAGs.
-The nodes of the DAG represent data frames to be produced, while the edges of the DAG represent the operations need to construct the data frames.
-This is inverted from the Query DAG where the nodes are operations and edges represents data sets.
+## Planning and Optmimization
+-----------------------------
 
-The leaves of the plan DAG represent sources of data and the data flows from bottom up through the tree.
-Again this is inverted from the Query DAG where data flows top down.
+```go
+type LogicalPlanner interface {
+	Plan(*flux.Spec) (*PlanSpec, error)
+}
 
-## Data Frames
+type PhysicalPlanner interface {
+	Plan(*PlanSpec) (*PlanSpec, error)
```

Comment:
Practically, today it's not possible to use just the logical planner and skip physical planning, because a each `from` needs a `range` to be pushed into it, and that push down is done as a physical optimization.

Some more depth: right now all our procedure specs are both logical procedure specs and physical procedure specs. That is, they all have a `Cost` method and can be translated directly into an instance of `Transformation` or `Source`.  But that won't always be the case; at some point we'll have (for example) a `JoinProcedureSpec` which is logical (no `Cost` method and can't be converted to a `Transformation`) and the physical planner will need to choose either `MergeJoinProcedureSpec` or `NestedLoopProcedureSpec` for the physical plan.

(I need to add the above to the docs.)

I think you're right that the logical planner should return something more opaque.  Josh and I have talked about breaking up `PlanSpec` into logical and physical counterparts.

As to why the logical/physical divide is exposed on the interface level, that's a good question.  Honestly, I was following the example of what was there in the `control` package already.  I can think of good reasons for the divide, like maybe wanting to cache logical plans to use later on, but that seems to be a ways off.

Reply:
👍 I like the answer and I'm really looking forward to a further API-level differentiation between the two. I can easily see now why the difference would be needed, but we do need to have them as separate APIs. The influxql query engine suffered for a long time because the outward facing  APIs made a lot of assumptions about the previous steps that were invoked without having them encoded as part of the type system so it's important that we do that before having a stable API.